### PR TITLE
Test for two factor authentication

### PR DESCRIPTION
**Pull request by bhawanaprasain**
## Description
The result of last verification request made using OTP can be checked via API.

## Related Issue
 #38

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [X] Tests only (no source changes)

### CODE REVIEW

**Comment by bhawanaprasain** [comment]
@dpakach @skshetry @paurakhsharma please review on my PR.

File: tests/acceptance/features/bootstrap/TwoFactorTOTPContext.php [hunk 1]
```diff
@@ -250,6 +250,104 @@ public function theSecretCodeFromQRCodeShouldMatchWithTheOneDisplayedOnTheWebUI(
...
+	/**
+	 * Send request with secet key for two factor authentication
+	 *
+	 * @param string $user
+	 * @param string $secretKey
+	 *
+	 * @return void
+	 */
+	public function sendRequestWithSecretKey($user, $secretKey) {
+		$response = \TestHelpers\OcsApiHelper::sendRequest(
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/TwoFactorTOTPContext.php]: import `OcsApiHelper` into the namespace

File: tests/acceptance/features/bootstrap/TwoFactorTOTPContext.php [hunk 2]
```diff
@@ -250,6 +250,104 @@ public function theSecretCodeFromQRCodeShouldMatchWithTheOneDisplayedOnTheWebUI(
...
+			'GET',
+			"/apps/twofactor_totp/api/v1/validate/$user/$secretKey",
+			$body = [],
+			$ocsApiVersion = 1
+		);
+		$this->featureContext->setResponse($response);
+	}
+
+	/**
+	 * @When the administrator tries to verify the otp key for user :user using the correct key
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/TwoFactorTOTPContext.php]:
```suggestion
	 * @When the administrator tries to verify the one-time key generated from the secret key for user :user
```

File: tests/acceptance/features/apiTwoFactorTOTP/steps.feature [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+Feature: Testing Two factor TOTP
+  As a admin
+  I want to be able to verify secrets
+  So that the users can use TOTP without verification with TOTP code
+
+  Background:
+    Given these users have been created with default attributes and skeleton files:
+      | username |
+      | user0    |
+    And using OCS API version "1"
```
Comment (skshetry) [inline tests/acceptance/features/apiTwoFactorTOTP/steps.feature]: Can you also check for both OCS versions?

File: tests/acceptance/features/apiTwoFactorTOTP/steps.feature [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+  Scenario: Administrator tries to verify OTP key for user using correct key
+    Given user "user0" has logged in using the webUI
+    And the user has browsed to the personal security settings page
+    And the user has activated TOTP Second-factor auth but not verified
+    When the administrator tries to verify the otp key for user "user0" using the correct key
+    Then the OCS status code should be "100"
```
Comment (skshetry) [inline tests/acceptance/features/apiTwoFactorTOTP/steps.feature]: Also, can you check for Http status code

File: tests/acceptance/features/bootstrap/TwoFactorTOTPContext.php [hunk 5]
```diff
@@ -250,6 +250,104 @@ public function theSecretCodeFromQRCodeShouldMatchWithTheOneDisplayedOnTheWebUI(
...
+		)['result'];
+		if ((int)$result !== 1) {
+			throw new \Exception(
+				"Invalid OTP"
+			);
+		}
+	}
+
+	/**
+	 * @When the administrator tries to verify the otp key for user :user using the wrong key
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/TwoFactorTOTPContext.php]:
```suggestion
	 * @When the administrator tries to verify with the random key :key for user :user
```

Comment (skshetry) [reply tests/acceptance/features/bootstrap/TwoFactorTOTPContext.php]: Or, maybe this:
```suggestion
	 * @When the administrator tries to verify with the random key for user :bust_in_silhouette: 
```
I do prefer the first one though.

File: tests/acceptance/features/bootstrap/TwoFactorTOTPContext.php [hunk 6]
```diff
@@ -250,6 +250,104 @@ public function theSecretCodeFromQRCodeShouldMatchWithTheOneDisplayedOnTheWebUI(
...
+			);
+		}
+	}
+
+	/**
+	 * @When the administrator tries to verify the otp key for a user that does not exist
+	 *
+	 * @return void
+	 */
+	public function theAdministratorTriesToVerifyTheOtpKeyForUserAUserThatDoesNotExist() {
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/TwoFactorTOTPContext.php]: Do you need this step? You could achieve the same thing with above steps as well

File: tests/acceptance/features/bootstrap/TwoFactorTOTPContext.php [hunk 7]
```diff
@@ -250,6 +251,93 @@ public function theSecretCodeFromQRCodeShouldMatchWithTheOneDisplayedOnTheWebUI(
...
+	}
+
+	/**
+	 * @When the administrator tries to verify with the random key :key for user :user
+	 *
+	 * @param string $user
+	 *
+	 * @return void
+	 */
+	public function theAdministratorTriesToVerifyTheOtpKeyForUserUsingTheWrongKey($user) {
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/TwoFactorTOTPContext.php]:
```suggestion
	public function theAdministratorTriesToVerifyTheOtpKeyForUserUsingTheWrongKey($user, $key) {
```
Also, you need to change docs, and the body of the function to use the `Skey`.

File: tests/acceptance/features/apiTwoFactorTOTP/steps.feature [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    When the administrator tries to verify with the one-time key generated from the secret key for user "user0"
+    Then the OCS status code should be "100"
+    And the HTTP status code should be "200"
+    And the result of the last verification request should be true
+
+  Scenario: Administrator tries to verify OTP key for user using wrong key
+    Given user "user0" has logged in using the webUI
+    And the user has browsed to the personal security settings page
+    And the user has activated TOTP Second-factor auth but not verified
+    When the administrator tries to verify with the random key "random" for user "user0"
```
Comment (individual-it) [inline tests/acceptance/features/apiTwoFactorTOTP/steps.feature]:
```suggestion
    When the administrator tries to verify with the invalid key "random" for user "user0"
```

**Comment by bhawanaprasain** [comment]
@individual-it @phil-davis please review again

**Pull request merged**